Cast-align warnings in latest compiler version (gcc 12)

**Describe the bug**
When compiling with GCC version 12.2, along with the `-Wcast-align=strict` option, some cast alignment warnings appear.

For example:
```
/home/joe/code/cfecfs/github/libs/bplib/mpool/src/v7_mpool.c: In function ‘bplib_mpool_get_block_content’:
/home/joe/code/cfecfs/github/libs/bplib/mpool/src/v7_mpool.c:169:16: error: cast increases required alignment of target type [-Werror=cast-align]
  169 |         return (bplib_mpool_block_content_t *)cb;
      |                ^
```

**To Reproduce**
Build with GCC 12.x and use `-Wcast-align=strict` compiler option.

**Expected behavior**
Should build clean.

**System observed on:**
Anything with latest GCC and/or strict alignment requirements.

**Additional context**
These are all pointer arithmetic operations, generally from taking a base pointer and adding the offset to an internal object within.  They are safe because the pointer address offset being used in the calculation was originally derived using `offsetof()` with the correct struct definition, so this will be correctly aligned.

In general the warning can be squelched by passing it through the `void*` type.

**Reporter Info**
Joseph Hickey, Vantage Systems, Inc.

--- a/cache/src/v7_cache_internal.h
+++ b/cache/src/v7_cache_internal.h
@@ -185,7 +185,7 @@ static inline bplib_mpool_flow_t *bplib_cache_get_flow(bplib_cache_state_t *stat
     bplib_cache_entry_get_container(ptr, offsetof(bplib_cache_entry_t, member))
 static inline bplib_cache_entry_t *bplib_cache_entry_get_container(const bplib_rbt_link_t *link, size_t offset)
 {
-    return (bplib_cache_entry_t *)((uint8_t *)link - offset);
+    return (bplib_cache_entry_t *)(void *)((uint8_t *)link - offset);
 }
 
 void bplib_cache_custody_finalize_dacs(bplib_cache_state_t *state, bplib_cache_entry_t *store_entry);

--- a/lib/src/v7_routing.c
+++ b/lib/src/v7_routing.c
@@ -337,7 +337,7 @@ int bplib_route_register_handler_impl(bplib_routetbl_t *tbl, bp_handle_t intf_id
 
     base_ptr = (uint8_t *)ifp;
 
-    subq = (bplib_mpool_subq_workitem_t *)(base_ptr + func_position);
+    subq = (bplib_mpool_subq_workitem_t *)(void *)(base_ptr + func_position);
 
     if (subq->job_header.handler == new_func)
     {

--- a/mpool/inc/v7_mpool.h
+++ b/mpool/inc/v7_mpool.h
@@ -256,7 +256,7 @@ static inline bool bplib_mpool_is_indirect_block(const bplib_mpool_block_t *cb)
  */
 static inline bool bplib_mpool_is_any_content_node(const bplib_mpool_block_t *cb)
 {
-    return (cb->type >= bplib_mpool_blocktype_generic && cb->type < bplib_mpool_blocktype_max);
+    return (cb->type > bplib_mpool_blocktype_undefined && cb->type < bplib_mpool_blocktype_max);
 }
 
 /**

--- a/mpool/src/v7_mpool.c
+++ b/mpool/src/v7_mpool.c
@@ -151,7 +151,7 @@ bplib_mpool_block_t *bplib_mpool_get_block_from_link(bplib_mpool_block_t *lblk)
     while (bblk != NULL && bplib_mpool_is_secondary_index_node(bblk) && bblk->parent_offset != 0)
     {
         /* the parent_offset field indicates this block position within the parent */
-        bblk = (bplib_mpool_block_t *)((uint8_t *)bblk - bblk->parent_offset);
+        bblk = (bplib_mpool_block_t *)(void *)((uint8_t *)bblk - bblk->parent_offset);
     }
 
     return bblk;
@@ -166,7 +166,7 @@ bplib_mpool_block_content_t *bplib_mpool_get_block_content(bplib_mpool_block_t *
 {
     if (cb != NULL && bplib_mpool_is_any_content_node(cb))
     {
-        return (bplib_mpool_block_content_t *)cb;
+        return (bplib_mpool_block_content_t *)(void *)cb;
     }
     return NULL;
 }
@@ -180,7 +180,7 @@ const bplib_mpool_block_content_t *bplib_mpool_get_block_content_const(const bpl
 {
     if (cb != NULL && bplib_mpool_is_any_content_node(cb))
     {
-        return (const bplib_mpool_block_content_t *)cb;
+        return (const bplib_mpool_block_content_t *)(const void *)cb;
     }
     return NULL;
 }
@@ -431,7 +431,7 @@ void *bplib_mpool_generic_data_cast(bplib_mpool_block_t *cb, uint32_t required_m
      * associated with it, not just a generic_data block.  The difference is that the generic
      * data block _only_ has the generic data, whereas the other block types can have both. */
     result = NULL;
-    block  = (bplib_mpool_block_content_t *)cb;
+    block  = bplib_mpool_get_block_content(cb);
     while (block != NULL && bplib_mpool_is_any_content_node(&block->header.base_link))
     {
         if (block->header.content_type_signature == required_magic)
@@ -473,7 +473,7 @@ bplib_mpool_block_t *bplib_mpool_generic_data_uncast(void *blk, bplib_mpool_bloc
     }
 
     data_offset += offsetof(bplib_mpool_block_content_t, u);
-    block = (bplib_mpool_block_content_t *)((uint8_t *)blk - data_offset);
+    block = (bplib_mpool_block_content_t *)(void *)((uint8_t *)blk - data_offset);
     if (block->header.base_link.type != parent_bt || block->header.content_type_signature != required_magic)
     {
         return NULL;
@@ -542,8 +542,8 @@ bplib_mpool_block_content_t *bplib_mpool_alloc_block_internal(bplib_mpool_t *poo
     }
 
     /* figure out how to initialize this block by looking up the content type */
-    api_block =
-        (bplib_mpool_api_content_t *)bplib_rbt_search_unique(content_type_signature, &admin->blocktype_registry);
+    api_block = (bplib_mpool_api_content_t *)(void *)bplib_rbt_search_unique(content_type_signature,
+                                                                             &admin->blocktype_registry);
     if (api_block == NULL)
     {
         /* no constructor, cannot create the block! */
@@ -579,7 +579,8 @@ bplib_mpool_block_content_t *bplib_mpool_alloc_block_internal(bplib_mpool_t *poo
         admin->max_alloc_watermark = block_count;
     }
 
-    block = (bplib_mpool_block_content_t *)node;
+    node->type = blocktype;
+    block      = bplib_mpool_get_block_content(node);
 
     /*
      * zero fill the content part first, this ensures that this is always done,
@@ -604,8 +605,6 @@ bplib_mpool_block_content_t *bplib_mpool_alloc_block_internal(bplib_mpool_t *poo
             /* nothing more for this node type (this catches cbor_data)  */
             break;
     }
-
-    node->type = blocktype;
 
     /* If the module did supply a constructor, invoke it now */
     if (api_block->api.construct != NULL)
@@ -937,13 +936,13 @@ uint32_t bplib_mpool_collect_blocks(bplib_mpool_t *pool, uint32_t limit)
 
         /* recycled blocks must all be "real" blocks (not secondary refs or head nodes, etc) and
          * have refcount of 0, or else bad things might happen */
-        assert(bplib_mpool_is_any_content_node(rblk));
-        content = (bplib_mpool_block_content_t *)rblk;
+        content = bplib_mpool_get_block_content(rblk);
+        assert(content != NULL);
         assert(content->header.refcount == 0);
 
         /* figure out how to de-initialize the user content by looking up the content type */
-        api_block = (bplib_mpool_api_content_t *)bplib_rbt_search_unique(content->header.content_type_signature,
-                                                                         &admin->blocktype_registry);
+        api_block = (bplib_mpool_api_content_t *)(void *)bplib_rbt_search_unique(content->header.content_type_signature,
+                                                                                 &admin->blocktype_registry);
 
         if (api_block != NULL)
         {
@@ -1086,9 +1085,9 @@ void bplib_mpool_debug_print_list_stats(bplib_mpool_block_t *list, const char *l
             break;
         }
 
-        if (bplib_mpool_is_any_content_node(blk))
+        content = bplib_mpool_get_block_content(blk);
+        if (content != NULL)
         {
-            content = (bplib_mpool_block_content_t *)blk;
             printf("DEBUG: %s(): block addr=%lx type=%d refcount=%u\n", __func__, (unsigned long)blk,
                    content->header.base_link.type, content->header.refcount);
 

--- a/mpool/src/v7_mpool_ref.c
+++ b/mpool/src/v7_mpool_ref.c
@@ -98,7 +98,8 @@ bplib_mpool_block_t *bplib_mpool_ref_make_block(bplib_mpool_ref_t refptr, uint32
     pool = bplib_mpool_get_parent_pool_from_link(&bblk->header.base_link);
 
     lock = bplib_mpool_lock_resource(pool);
-    rblk = bplib_mpool_alloc_block_internal(pool, bplib_mpool_blocktype_ref, magic_number, init_arg, BPLIB_MPOOL_ALLOC_PRI_MHI);
+    rblk = bplib_mpool_alloc_block_internal(pool, bplib_mpool_blocktype_ref, magic_number, init_arg,
+                                            BPLIB_MPOOL_ALLOC_PRI_MHI);
     bplib_mpool_lock_release(lock);
 
     if (rblk == NULL)
@@ -125,7 +126,7 @@ bplib_mpool_ref_t bplib_mpool_ref_from_block(bplib_mpool_block_t *rblk)
         return NULL;
     }
 
-    content = (bplib_mpool_block_content_t *)rblk;
+    content = bplib_mpool_get_block_content(rblk);
 
     return bplib_mpool_ref_duplicate(content->u.ref.pref_target);
 }

